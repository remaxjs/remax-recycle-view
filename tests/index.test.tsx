/* eslint-disable global-require */
import React from 'react';
import renderer from 'react-test-renderer';

describe('src/index.tsx', () => {
  it(`should render correct LIST_HEIGHT 0rpx when no data`, () => {
    jest.isolateModules(() => {
      const RecycleView = require('../src/index').default;
      const component = renderer.create(<RecycleView renderItem={() => <div>hello</div>} />);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it(`should render correct LIST_HEIGHT 200rpx`, () => {
    jest.isolateModules(() => {
      const RecycleView = require('../src/index').default;
      const component = renderer.create(
        <RecycleView
          data={[{ height: 100 }, { height: 100 }]}
          renderItem={() => <div>hello</div>}
        />,
      );
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it('should render placeholder-image, header and bottom', () => {
    jest.isolateModules(() => {
      const RecycleView = require('../src/index').default;
      const component = renderer.create(
        <RecycleView
          headerHeight={50}
          bottomHeight={50}
          renderHeader={() => <div>header</div>}
          renderBottom={() => <div>bottom</div>}
          placeholderImage="https://via.placeholder.com/150"
          data={[{ height: 100 }]}
          renderItem={() => <div>hello</div>}
        />,
      );
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it('should render correct items when handle scroll', () => {
    jest.isolateModules(() => {
      const RecycleView = require('../src/index').default;
      const { ScrollView } = require('../src/remax');
      const mockData = [];
      let testNumber = 0;
      for (let i = 0; i < 500; i++) {
        mockData.push({
          height: 100,
          text: i,
        });
      }
      const component = renderer.create(
        <RecycleView
          className="recycle-view"
          onScroll={() => {
            testNumber++;
          }}
          data={mockData}
          overscanCount={5}
          headerHeight={100}
          bottomHeight={100}
          renderHeader={() => {
            return <div style={{ height: 100, backgroundColor: 'yellow' }}>this is header</div>;
          }}
          renderBottom={() => {
            return <div style={{ height: 100, backgroundColor: 'yellow' }}>this is bottom</div>;
          }}
          renderItem={({ text }) => {
            return <div style={{ height: '100%' }}>{text}</div>;
          }}
        />,
      );
      const instance = component.root;
      const recycleView = instance.findByType(ScrollView);

      const itemCount = (component.toJSON().children[1] as any).children[0].children.length;
      expect(itemCount).toBe(11);

      // mock scroll to middle
      renderer.act(() => {
        const handleScroll = recycleView.props.onScroll;
        handleScroll({ detail: { scrollTop: 25150, scrollHeight: 50200 } });
      });

      const items = (component.toJSON().children[1] as any).children[0].children;
      const itemCountAfterScroll = items.length;
      expect(itemCountAfterScroll).toBe(16);
      expect(items[0].children[0].children[0]).toBe('245');
      expect(testNumber).toBe(1);
    });
  });
});
