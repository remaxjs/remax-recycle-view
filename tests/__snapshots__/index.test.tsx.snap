// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/index.tsx should render correct LIST_HEIGHT 200rpx 1`] = `
<scroll-view
  onScroll={[Function]}
  scrollY={true}
>
  <view
    style={
      Object {
        "background": undefined,
        "backgroundSize": undefined,
        "height": 200,
        "position": "relative",
      }
    }
  >
    <view
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <view
        className="recycle-view-item"
        style={
          Object {
            "height": 100,
          }
        }
      >
        <div>
          hello
        </div>
      </view>
      <view
        className="recycle-view-item"
        style={
          Object {
            "height": 100,
          }
        }
      >
        <div>
          hello
        </div>
      </view>
    </view>
  </view>
</scroll-view>
`;

exports[`src/index.tsx should render correctly in platform ali 1`] = `
<scroll-view
  onScroll={[Function]}
  scrollY={true}
>
  <view
    style={
      Object {
        "background": undefined,
        "backgroundSize": undefined,
        "height": 0,
        "position": "relative",
      }
    }
  >
    <view
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    />
  </view>
</scroll-view>
`;

exports[`src/index.tsx should render correctly in platform toutiao 1`] = `
<scroll-view
  onScroll={[Function]}
  scrollY={true}
>
  <view
    style={
      Object {
        "background": undefined,
        "backgroundSize": undefined,
        "height": 0,
        "position": "relative",
      }
    }
  >
    <view
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    />
  </view>
</scroll-view>
`;

exports[`src/index.tsx should render correctly in platform wechat 1`] = `
<scroll-view
  lowerThreshold={50}
  onScroll={[Function]}
  scrollY={true}
  upperThreshold={50}
>
  <view
    style={
      Object {
        "background": undefined,
        "backgroundSize": undefined,
        "height": 0,
        "position": "relative",
      }
    }
  >
    <view
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    />
  </view>
</scroll-view>
`;

exports[`src/index.tsx should render placeholder-image, header and bottom 1`] = `
<scroll-view
  onScroll={[Function]}
  scrollY={true}
>
  <view
    className="recycle-view-header"
    style={
      Object {
        "height": 50,
      }
    }
  >
    <div>
      header
    </div>
  </view>
  <view
    style={
      Object {
        "background": "url(\\"https://via.placeholder.com/150\\") repeat-y",
        "backgroundSize": "contain",
        "height": 100,
        "position": "relative",
      }
    }
  >
    <view
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <view
        className="recycle-view-item"
        style={
          Object {
            "height": 100,
          }
        }
      >
        <div>
          hello
        </div>
      </view>
    </view>
  </view>
  <view
    className="recycle-view-bottom"
    style={
      Object {
        "height": 50,
      }
    }
  >
    <div>
      bottom
    </div>
  </view>
</scroll-view>
`;
